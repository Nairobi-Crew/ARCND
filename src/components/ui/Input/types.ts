import { FC, InputHTMLAttributes } from 'react';

export type TOnValueChange = (value: string) => void;

export type OwnInputProps = {
  id: string
  label?: string
  type?: string
  labelClasses?: string[]
  inputClasses?: string[]
  onValueChanged?: TOnValueChange
} & InputHTMLAttributes<HTMLInputElement>;

export type InputProps = FC<OwnInputProps>;
