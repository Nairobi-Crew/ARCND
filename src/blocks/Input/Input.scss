@import '../../common/styles/common.scss';

.input {
  font-weight: 400;
  height: 56px;
  width: 100%;
  background-color: inherit;
  color: #e2e0e0;

  &__label {
    position: relative;
    display: block;
    height: 100%;
    width: 100%;
    background-color: inherit;
  }

  &__input {
    height: 100%;
    width: 100%;
    outline: none;
    border: none;
    border-bottom: 1px solid #464545;
    background-color: inherit;
    color: #e2e0e0;
  }

  &__input:focus {
    border-color: $violet;
  }

  &__input:invalid,
  &__input._error {
    border-color: $red;
  }

  &__input:valid:not(:placeholder-shown):not(:focus):not(._error) {
    border-color: $aqua;
  }

  &__input::placeholder {
    opacity: 0;
  }

  &__label-text {
    position: absolute;
    font-size: 14px;
    top: 50%;
    left: 0;
    padding: 2px 8px 2px 2px;
    background-color: inherit;
    transform: translateY(-50%);
    transition: all 0.3s;
  }

  &__label-text > sup {
    position: absolute;
    bottom: 5px;
    right: 2px;
    color: $red;
  }

  &__input:not(:placeholder-shown) + &__label-text,
  &__input:focus + &__label-text {
    top: 0;
    font-size: 9px;
  }

  &__error-message {
    font-size: 12px;
    color: $red;
  }
}
