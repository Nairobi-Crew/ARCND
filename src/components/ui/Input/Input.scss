@import "src/common/utils";

.input {
  font-weight: 400;
  height: 56px;
  width: 100%;
  background-color: inherit;
  @include font-color;
  margin: 0 0 40px;

  textarea {
    padding-top: 12px;
  }

  &__label {
    position: relative;
    display: block;
    height: 100%;
    width: 100%;
    font-family: $ff;
    @include font-color;
    background-color: inherit;
  }

  &__input {
    resize: none;
    height: 100%;
    width: 100%;
    outline: none;
    border: none;
    border-bottom: 2px solid #49005d;
    background-color: inherit;
    font-family: $ff;
    @include font-color;
  }

  &__input:focus {
    border-color: $primary-btn;
  }

  &__input:invalid:not(:placeholder-shown),
  &__input_error {
    border-color: $error;
  }

  &__input:valid:not(:placeholder-shown):not(:focus):not(&__input_error) {
    border-color: $secondary-btn;
  }

  &__input::placeholder {
    opacity: 0;
  }

  &__label-text {
    position: absolute;
    font-size: $f-m;
    font-family: $ff;
    top: 50%;
    left: 0;
    padding: 2px 8px 2px 2px;
    background-color: inherit;
    transform: translateY(-50%);
    transition: all 0.3s;
  }

  &__label-text > sup {
    position: absolute;
    bottom: 5px;
    right: 2px;
    color: $error;
  }

  &__input:not(:placeholder-shown) + &__label-text,
  &__input:focus + &__label-text {
    top: 0;
    font-size: 9px;
  }

  &__error-message {
    font-size: $f-m;
    color: $error;
  }
}
