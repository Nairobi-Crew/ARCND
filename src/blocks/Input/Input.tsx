import React from "react";
import '../../common/styles/common.scss'

import './Input.scss'

interface IInputProps {
  type?: string;
  label?: string;
  errorMessage?: string;
  name?: string;
  value?: string;
}

export default function Input({type, label, errorMessage, name, value}:IInputProps) {
  return (
    <div className="input">
      <label className="input__label">
        <input type={type}
               className="input__input"
               name={name}
               value={value}
               placeholder={label}/>
          <span className="input__label-text">{label}</span>
      </label>
      <span className="input__error-message">{errorMessage}</span>
    </div>
  )
}
