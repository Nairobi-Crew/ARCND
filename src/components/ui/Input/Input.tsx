import React, { useEffect, useState } from 'react';
import { InputProps } from 'UI/Input/types';
import './Input.scss';

const Input: InputProps = ({
  label, value, onValueChanged, errorMessage, ...restProps
}) => {
  const [inputValue, setInputValue] = useState<string>(value as string || '');
  useEffect(() => {
    setInputValue(value as string);
  }, [value]);
  const onInputHandler = (e: React.FormEvent<HTMLInputElement>): void => {
    if (e.target) {
      const v = (e.target as HTMLInputElement).value;
      setInputValue(v);
      if (onValueChanged) {
        onValueChanged(v);
      }
    }
  };

  return (
    <div className="input">
      <label className="input__label">
        <input
          className="input__input"
          value={inputValue}
          placeholder={label}
          onInput={onInputHandler}
          {...restProps}
        />
        <span className="input__label-text">{label}</span>
      </label>
      <span className="input__error-message">{errorMessage}</span>
    </div>
  );
};

export default Input;
