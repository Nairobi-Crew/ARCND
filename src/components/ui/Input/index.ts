import Input from 'UI/Input/Input';
import { memo } from 'react';
import { OwnInputProps } from 'UI/Input/types';

export default memo<OwnInputProps>(Input);
