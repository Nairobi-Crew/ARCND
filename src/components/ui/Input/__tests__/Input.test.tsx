/* eslint-disable no-undef */
import renderer from 'react-test-renderer';
import { mount } from 'enzyme';
import React from 'react';
import Input from 'UI/Input/Input';

describe('Render Input', () => {
  it('Render correct', () => {
    const Component = () => (
      <Input
        errorMessage="Test error message"
        disabled={false}
        value="TestValue"
        onChange={(e) => e.target.value}
        onValueChanged={(v) => v}
      />
    );
    const inputComponent = renderer.create(<Component />).toJSON();
    expect(inputComponent).toMatchSnapshot();
  });

  it('value="test val1"', () => {
    const wrapper = mount(<Input value="test val1" />);
    expect(wrapper.prop('value')).toEqual('test val1');
    expect(wrapper.prop('value')).not.toEqual('test val1 ');
    expect(wrapper.prop('value')).not.toEqual('Test val1');
  });

  it('errorMessage=""', () => {
    const wrapper = mount(<Input errorMessage="" value="Nothing" />);
    expect(wrapper.find('.input__error-message').at(0).text()).toEqual('');
    expect(wrapper.prop('errorMessage')).toEqual('');
  });

  const message = 'TeSt Me$$a9e';
  it(`errorMessage="${message}"`, () => {
    const wrapper = mount(<Input errorMessage={message} value="Nothing" />);
    expect(wrapper.find('.input__error-message').at(0).text()).toEqual(message);
    expect(wrapper.prop('errorMessage')).toEqual(message);
  });
});
