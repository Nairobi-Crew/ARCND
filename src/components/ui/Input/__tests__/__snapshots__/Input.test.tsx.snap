// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Render Textarea Render correct 1`] = `
<div
  className="input"
>
  <label
    className="input__label"
  >
    <input
      className="input__input"
      disabled={false}
      onChange={[Function]}
      onInput={[Function]}
      value="TestValue"
    />
    <span
      className="input__label-text"
    />
  </label>
  <span
    className="input__error-message"
  >
    Test error message
  </span>
</div>
`;
